rsa/pkcs1_v2.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+#
+#  Copyright 2011 Sybren A. Stüvel <sybren@stuvel.eu>
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""Functions for PKCS#1 version 2 encryption and signing
+
+This module implements certain functionality from PKCS#1 version 2. Main
+documentation is RFC 2437: https://tools.ietf.org/html/rfc2437
+"""
+
+from rsa._compat import range
+from rsa import (
+    common,
+    pkcs1,
+    transform,
+)
+
+HASH_METHOD_TO_BYTE_LENGTH = {
+    'MD5': 16,
+    'SHA-1': 20,
+    'SHA-256': 28,
+    'SHA-384': 48,
+    'SHA-512': 64,
+}
+
+
+def mgf1(seed, length, hasher='SHA-1'):
+    """
+    MGF1 is a Mask Generation Function based on a hash function.
+
+    A mask generation function takes an octet string of variable length and a
+    desired output length as input, and outputs an octet string of the desired
+    length. The plaintext-awareness of RSAES-OAEP relies on the random nature of
+    the output of the mask generation function, which in turn relies on the
+    random nature of the underlying hash.
+
+    :param seed: seed from which mask is generated, an octet string
+    :param length: intended length in octets of the mask, at most 2^32(hLen)
+    :param hasher: hash function (hLen denotes the length in octets of the hash
+        function output)
+
+    :return: mask, an octet string of length `length`
+
+    :raise OverflowError: when `length` is too large for the specified `hasher`
+    :raise ValueError: when specified `hasher` is invalid
+    """
+
+    try:
+        hash_length = HASH_METHOD_TO_BYTE_LENGTH[hasher]
+    except KeyError:
+        raise ValueError(
+            'Invalid `hasher` specified. Please select one of: {hash_list}'.format(
+                hash_list=', '.join(sorted(HASH_METHOD_TO_BYTE_LENGTH.keys()))
+            )
+        )
+
+    # If l > 2^32(hLen), output "mask too long" and stop.
+    if length > (2**32 * hash_length):
+        raise OverflowError
+
+    # Let output be the empty octet string.
+    output = b''
+
+    # For counter from 0 to ceil(l / hLen)-1, do the following:
+    for counter in range(common.ceil_div(length, hash_length) + 1):
+        # Convert counter to an octet string C of length 4 with the primitive
+        # I2OSP
+        c = transform.int2bytes(counter, fill_size=4)
+
+        # Concatenate the hash of the `seed` and C to the octet string `output`
+        output += pkcs1._hash(seed + c, method_name=hasher)
```

Comment:
This is a bad idea. `output` is stored using a fixed-length array, and concatenating to that will cause a copy to be made. Instead, append each output of `pkcs1._hash()` to a list, then return `b''.join(thelist)`.